### PR TITLE
mgr: handle race with finisher after shutdown

### PR DESCRIPTION
**Pull request by batrick**

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test dashboard

**Comment by batrick** [comment]
jenkins test dashboard

**Comment by batrick** [comment]
I was thrashing some changes to the mdsmap/osdmap and that seemed to trigger it at least once:

````
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown module volumes shutdown() returned
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joining module balancer
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joined module balancer
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joining module crash
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joined module crash
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joining module devicehealth
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joined module devicehealth
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joining module iostat
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joined module iostat
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joining module orchestrator_cli
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joined module orchestrator_cli
2019-11-13T19:49:12.281+0000 7f4398ee6700 10 mgr shutdown joining module pg_autoscaler
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joined module pg_autoscaler
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joining module progress
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joined module progress
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joining module rbd_support
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joined module rbd_support
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joining module restful
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joined module restful
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joining module status
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joined module status
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joining module telemetry
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joined module telemetry
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joining module volumes
2019-11-13T19:49:12.282+0000 7f4398ee6700 10 mgr shutdown joined module volumes
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify mon_status
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify health
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700  5 mgr notify cancelling notify pg_summary
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr Gil Switched to new thread state 0x5647c9fe8000
2019-11-13T19:49:12.282+0000 7f4398ee6700 20 mgr ~Gil Destroying new thread state 0x5647c9fe8000
````

So looks like this fix should work.

File: src/mgr/ActivePyModules.cc [hunk 1]
```diff
@@ -1007,7 +1004,7 @@ void ActivePyModules::set_uri(const std::string& module_name,
 
   dout(4) << " module " << module_name << " set URI '" << uri << "'" << dendl;
 
-  modules[module_name]->set_uri(uri);
+  modules.at(module_name)->set_uri(uri);
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/mgr/ActivePyModules.cc]: It would segfault if that ever happened as it was a `std::unique_ptr` before which would initialize to `nullptr`.

**Comment by bk201** [comment]
jenkins test dashboard backend

**Comment by s0nea** [comment]
> jenkins test dashboard backend

The backend test failed because of:

```
2019-11-14 05:37:55,247.247 INFO:__main__:ERROR: test_create_export (tasks.mgr.dashboard.test_ganesha.GaneshaTest)
2019-11-14 05:37:55,247.247 INFO:__main__:----------------------------------------------------------------------
2019-11-14 05:37:55,247.247 INFO:__main__:Traceback (most recent call last):
2019-11-14 05:37:55,247.247 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 154, in setUp
2019-11-14 05:37:55,248.248 INFO:__main__:    self.wait_for_health_clear(20)
2019-11-14 05:37:55,248.248 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/ceph_test_case.py", line 131, in wait_for_health_clear
2019-11-14 05:37:55,248.248 INFO:__main__:    self.wait_until_true(is_clear, timeout)
2019-11-14 05:37:55,248.248 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/ceph_test_case.py", line 163, in wait_until_true
2019-11-14 05:37:55,249.249 INFO:__main__:    raise RuntimeError("Timed out after {0}s".format(elapsed))
2019-11-14 05:37:55,249.249 INFO:__main__:RuntimeError: Timed out after 20s
```

It will be fixed here: https://github.com/ceph/ceph/pull/31484/. I'll test the changes in my PR

**Comment by callithea** [comment]
I'm still facing errors like mentioned in the issue when running the dashboard backend API tests locally:
```2019-11-14 11:10:30,328.328 INFO:__main__:Found ps line for daemon: 45548 ?        Ssl    0:06 ./bin/./ceph-mgr -i x
2019-11-14 11:10:31,329.329 INFO:__main__:Running ['ps', 'ww', '-u0']
2019-11-14 11:10:31,348.348 INFO:__main__:Found ps line for daemon: 45548 ?        Ssl    0:06 ./bin/./ceph-mgr -i x
2019-11-14 11:10:32,350.350 INFO:__main__:Running ['ps', 'ww', '-u0']
2019-11-14 11:10:32,371.371 INFO:__main__:Found ps line for daemon: 45548 ?        Ssl    0:06 ./bin/./ceph-mgr -i x
2019-11-14 11:10:33,373.373 INFO:__main__:Running ['ps', 'ww', '-u0']
2019-11-14 11:10:33,395.395 INFO:__main__:Found ps line for daemon: 45548 ?        Ssl    0:06 ./bin/./ceph-mgr -i x
2019-11-14 11:10:34,396.396 INFO:__main__:Running ['ps', 'ww', '-u0']
2019-11-14 11:10:34,418.418 INFO:__main__:Found ps line for daemon: 45548 ?        Ssl    0:06 ./bin/./ceph-mgr -i x
2019-11-14 11:10:35,419.419 INFO:__main__:Running ['ps', 'ww', '-u0']
2019-11-14 11:10:35,446.446 INFO:__main__:Found ps line for daemon: 45548 ?        Ssl    0:06 ./bin/./ceph-mgr -i x
2019-11-14 11:10:35,449.449 INFO:__main__:test_standby (tasks.mgr.test_dashboard.TestDashboard) ... ERROR
2019-11-14 11:10:35,450.450 INFO:__main__:Stopped test: test_standby (tasks.mgr.test_dashboard.TestDashboard) in 309.927945s
2019-11-14 11:10:35,450.450 INFO:__main__:
2019-11-14 11:10:35,451.451 INFO:__main__:======================================================================
2019-11-14 11:10:35,451.451 INFO:__main__:ERROR: test_standby (tasks.mgr.test_dashboard.TestDashboard)
2019-11-14 11:10:35,451.451 INFO:__main__:----------------------------------------------------------------------
2019-11-14 11:10:35,451.451 INFO:__main__:Traceback (most recent call last):
2019-11-14 11:10:35,451.451 INFO:__main__:  File "/ceph/qa/tasks/mgr/test_dashboard.py", line 18, in setUp
2019-11-14 11:10:35,452.452 INFO:__main__:    self._assign_ports("dashboard", "ssl_server_port")
2019-11-14 11:10:35,452.452 INFO:__main__:  File "/ceph/qa/tasks/mgr/mgr_test_case.py", line 193, in _assign_ports
2019-11-14 11:10:35,452.452 INFO:__main__:    cls.mgr_cluster.mgr_stop(mgr_id)
2019-11-14 11:10:35,452.452 INFO:__main__:  File "/ceph/qa/tasks/mgr/mgr_test_case.py", line 28, in mgr_stop
2019-11-14 11:10:35,452.452 INFO:__main__:    self.mgr_daemons[mgr_id].stop()
2019-11-14 11:10:35,453.453 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 443, in stop
2019-11-14 11:10:35,453.453 INFO:__main__:    self.daemon_type, self.daemon_id))
2019-11-14 11:10:35,453.453 INFO:__main__:MaxWhileTries: Timed out waiting for daemon mgr.x
2019-11-14 11:10:35,453.453 INFO:__main__:
2019-11-14 11:10:35,453.453 INFO:__main__:----------------------------------------------------------------------
2019-11-14 11:10:35,454.454 INFO:__main__:Ran 1 test in 325.068s
2019-11-14 11:10:35,454.454 INFO:__main__:
2019-11-14 11:10:35,454.454 INFO:__main__:FAILED (errors=1)
2019-11-14 11:10:35,454.454 INFO:__main__:
2019-11-14 11:10:35,454.454 INFO:__main__:======================================================================
2019-11-14 11:10:35,455.455 INFO:__main__:ERROR: test_standby (tasks.mgr.test_dashboard.TestDashboard)
2019-11-14 11:10:35,455.455 INFO:__main__:----------------------------------------------------------------------
2019-11-14 11:10:35,455.455 INFO:__main__:Traceback (most recent call last):
2019-11-14 11:10:35,455.455 INFO:__main__:  File "/ceph/qa/tasks/mgr/test_dashboard.py", line 18, in setUp
2019-11-14 11:10:35,455.455 INFO:__main__:    self._assign_ports("dashboard", "ssl_server_port")
2019-11-14 11:10:35,456.456 INFO:__main__:  File "/ceph/qa/tasks/mgr/mgr_test_case.py", line 193, in _assign_ports
2019-11-14 11:10:35,456.456 INFO:__main__:    cls.mgr_cluster.mgr_stop(mgr_id)
2019-11-14 11:10:35,456.456 INFO:__main__:  File "/ceph/qa/tasks/mgr/mgr_test_case.py", line 28, in mgr_stop
2019-11-14 11:10:35,456.456 INFO:__main__:    self.mgr_daemons[mgr_id].stop()
2019-11-14 11:10:35,456.456 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 443, in stop
2019-11-14 11:10:35,457.457 INFO:__main__:    self.daemon_type, self.daemon_id))
2019-11-14 11:10:35,457.457 INFO:__main__:MaxWhileTries: Timed out waiting for daemon mgr.x
2019-11-14 11:10:35,457.457 INFO:__main__:
```

**Comment by liewegas** [comment]
This fixes the other half of the problem: https://github.com/ceph/ceph/pull/31685

**Pull request merged**